apps/studio/src/utils/sitemap.ts

```diff
@@ -0,0 +1,101 @@
+import type { IsomerSitemap } from "@opengovsg/isomer-components"
+import type { Resource } from "@prisma/client"
+
+type ResourceDto = Omit<
+  Resource,
+  "id" | "parentId" | "publishedVersionId" | "draftBlobId"
+> & {
+  id: string
+  parentId: string | null
+}
+
+const getSitemapTreeFromArray = (
+  resources: ResourceDto[],
+  parentId: string | null,
+  path: string,
+): IsomerSitemap[] | undefined => {
+  if (resources.length === 0) {
+    return undefined
+  }
+
+  const children = resources.filter((resource) => {
+    if (parentId === null) {
+      return resource.parentId === null && resource.permalink !== ""
+    }
+    return resource.parentId === parentId
+  })
+
+  // Filter out children that have both the Page and the Folder types and keep
+  // only the one with the Folder type, then reassign the ID to the ID of the
+  // resource with the Page type
+  return children
+    .filter(
+      (child) =>
+        !(
+          child.type === "Page" &&
+          children.some(
+            (otherChild) =>
+              otherChild.id !== child.id &&
+              otherChild.permalink === child.permalink &&
+              otherChild.type === "Folder",
+          )
+        ),
+    )
+    .map((resource) => {
+      const permalink = `${path}${resource.permalink}`
+      if (resource.type === "Folder") {
+        const idOfPage = children.find(
+          (child) =>
+            child.id !== resource.id &&
+            child.permalink === resource.permalink &&
+            child.type === "Page",
+        )?.id
+
+        const folderChildren = resources.filter(
+          (item) => item.parentId === resource.id,
+        )
```

Comment:
do we not need another conditional here like `resource.type  === Folder`

Reply:
Hmm no, we are looking for all resources with that particular parentId, whereby the parentId is a resource that is a Folder. The subsequent recursion does the filtering between the Page and Folder.